File: Go/Perfect Number.go
// 0ms, 100.00%; 2.1MB, 100.00%
func checkPerfectNumber(num int) bool {
	if num < 6 {
		return false
	}
	target := num - 1
	for i, j := 2, int(math.Sqrt(float64(num))); i <= j; i++ {
		k := num / i
		if i*k != num {
			continue
		}
		target -= i + k
		if target < 0 || target > 0 && target < i+j {
			break
		}
	}
	return 0 == target
}

File: Java/Perfect Number.java
// 1ms, 95.17%; 35.9MB, 5.88%
class Solution {
    public static boolean checkPerfectNumber(int num) {
        if (num < 6) {
            return false;
        }
        int target = num - 1;
        for (int i = 2, j = (int) Math.sqrt(num); i <= j; ++i) {
            int k = num / i;
            if (i * k != num) {
                continue;
            }
            target -= i + k;
            if (target < 0 || target > 0 && target < i + j) {
                break;
            }
        }
        return 0 == target;
    }
}
